File: backend/app/schemas/inventory.py
from datetime import datetime

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Inventory Item schemas
# ---------------------------------------------------------------------------


class InventoryItemResponse(BaseModel):
    id: str
    company_id: str
    product_id: str
    product_name: str | None = None
    product_sku: str | None = None
    category_name: str | None = None
    quantity_on_hand: int
    reorder_point: int | None = None
    reorder_quantity: int | None = None
    location: str | None = None
    last_counted_at: datetime | None = None
    is_low_stock: bool = False
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PaginatedInventoryItems(BaseModel):
    items: list[InventoryItemResponse]
    total: int
    page: int
    per_page: int


class InventorySettingsUpdate(BaseModel):
    reorder_point: int | None = None
    reorder_quantity: int | None = None
    location: str | None = None


# ---------------------------------------------------------------------------
# Stock operation schemas
# ---------------------------------------------------------------------------


class ReceiveStockRequest(BaseModel):
    quantity: int = Field(..., gt=0, description="Quantity to receive (must be positive)")
    reference: str | None = None
    notes: str | None = None


class AdjustStockRequest(BaseModel):
    new_quantity: int = Field(..., ge=0, description="New stock quantity (must be >= 0)")
    reference: str | None = None
    notes: str | None = None


# ---------------------------------------------------------------------------
# Transaction schemas
# ---------------------------------------------------------------------------


class InventoryTransactionResponse(BaseModel):
    id: str
    product_id: str
    product_name: str | None = None
    transaction_type: str
    quantity_change: int
    quantity_after: int
    reference: str | None = None
    notes: str | None = None
    created_by: str | None = None
    created_by_name: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class PaginatedTransactions(BaseModel):
    items: list[InventoryTransactionResponse]
    total: int
    page: int
    per_page: int

File: frontend/src/types/inventory.ts
export interface InventoryItem {
  id: string;
  company_id: string;
  product_id: string;
  product_name: string | null;
  product_sku: string | null;
  category_name: string | null;
  quantity_on_hand: number;
  reorder_point: number | null;
  reorder_quantity: number | null;
  location: string | null;
  last_counted_at: string | null;
  is_low_stock: boolean;
  created_at: string;
  updated_at: string;
}

export interface PaginatedInventoryItems {
  items: InventoryItem[];
  total: number;
  page: number;
  per_page: number;
}

export interface ReceiveStockRequest {
  quantity: number;
  reference?: string;
  notes?: string;
}

export interface AdjustStockRequest {
  new_quantity: number;
  reference?: string;
  notes?: string;
}

export interface InventorySettingsUpdate {
  reorder_point?: number | null;
  reorder_quantity?: number | null;
  location?: string | null;
}

export interface ProductionEntryRequest {
  quantity: number;
  reference?: string;
  notes?: string;
}

export interface WriteOffRequest {
  quantity: number;
  reason: string;
  reference?: string;
  notes?: string;
}

export interface BatchProductionEntry {
  product_id: string;
  quantity: number;
  reference?: string;
  notes?: string;
}

export interface BatchProductionRequest {
  entries: BatchProductionEntry[];
  batch_reference?: string;
}

export interface BatchProductionResult {
  success_count: number;
  failure_count: number;
  results: Array<{
    product_id: string;
    success: boolean;
    error?: string;
  }>;
}

export interface InventoryTransaction {
  id: string;
  product_id: string;
  product_name: string | null;
  transaction_type: string;
  quantity_change: number;
  quantity_after: number;
  reference: string | null;
  notes: string | null;
  created_by: string | null;
  created_by_name: string | null;
  created_at: string;
}

export interface PaginatedTransactions {
  items: InventoryTransaction[];
  total: number;
  page: number;
  per_page: number;
}
